File: Week3/Code/get_TreeHeight.py
#!/bin/env python3

# Author: ph-u
# Script: get_TreeHeight.py
# Desc: R program substitution -- Tree Height calculation
# Input: python3 get_TreeHeight.py <.csv>
# Output: `<.csv>_treeheights.csv` in `Results` subdirectory
# Arguments: 1
# Date: Oct 2019

"""
R program substitution -- Tree Height calculation
"""

__appname__="get_TreeHeight.py"
__author__="ph-u"
__version__="0.0.1"
__license__="None"

## imports
import sys
import os
import math
import csv

## function
def TreeHeight(deg,dist):
    ## deg: degree, angle of elevation
    ## dist: distance between measurer and tree
    ## rad: radian, calculate from degree
    H=dist*math.tan(math.radians(deg))
    return H

## workflow

## if no sys.argv
if len(sys.argv)>1:

## working directory orientation issue
    os.chdir(os.path.dirname(sys.argv[1]))
    bn=os.path.basename(sys.argv[1]).split(".")[0]

## read csv
    f_0=list(csv.reader(open(sys.argv[1],"r")))
else:
    f_0=list(csv.reader(open("../Data/trees.csv","r")))
    bn="trees"

## calculation -- rubbish python core on data handling (https://stackoverflow.com/questions/44360162/how-to-access-a-column-in-a-list-of-lists-in-python)
for i in range(len(f_0)):
    if i==0: f_0[0].append("Tree.Height.m")
    else: f_0[i].append(TreeHeight(float(f_0[i][2]),float(f_0[i][1])))

## write result as csv
with open(str("../Results/"+bn+"_treeheights.csv"),"w") as csvfile: 
    csvW=csv.writer(csvfile)
    csvW.writerows(i for i in f_0)

File: Week3/Code/get_TreeHeight.R
#!/bin/env Rscript

# Author: PokMan Ho pok.ho19@imperial.ac.uk
# Script: get_TreeHeight.R
# Desc: Tree Height calculation program
# Input: Rscript get_TreeHeight.R
# Output: `<.csv>_treeheights.csv` in `Results` subdirectory
# Arguments: 1
# Date: Oct 2019

## This function calculates heights of trees given distance of each tree from its base and angle to its top, using the trigonometric formula

## height = distance * tan(radians)

## arguments
## degrees: The angle of elevation of tree
## distance: the distance from base of tree (e.g. metres)

## output
## the heights of the tree, same units as "distance"

TreeHeight <- function(degrees, distance){
  radians<-degrees*pi/180
  height<-distance*tan(radians)
  # print(paste("Tree Height is:",height))
  # 
  # return(height)
}
## TreeHeight(37,40)

## R homework
args=(commandArgs(T))
setwd(dirname(args[1]))
bn<-as.character(read.table(text=basename(args[1]),sep=".")[1,1])
# a<-read.csv("../Data/trees.csv",header = T)
a<-read.csv(args[1],header = T)
a.0<-data.frame(a,TreeHeight(a$Angle.degrees,a$Distance.m))
colnames(a.0)[dim(a.0)[2]]="Tree.Height.m"
# write.csv(a.0,"../Results/TreeHts.csv",quote = F,row.names = F)
write.csv(a.0,paste0("../Results/",bn,"_treeheights.csv"),quote = F,row.names = F)